File: julia/Euler/src/bin/p0082.jl

# project euler: problem 82

module Prob0082

function solve_0082(fname::String = "p082_matrix.txt", fn::Function = min)
    data = vcat(([parse(Int, x) for x in split(l, ",")]' for l in readlines(joinpath((@__DIR__), "../../assets", fname)))...)

    work = data[:, 1]
    for j = axes(data, 2)[begin + 1:end]
        crnt = data[:, j]
        work[begin] += crnt[begin]
        for i = axes(data, 1)[begin + 1:end]
            work[i] = crnt[i] + fn(work[i], work[i - 1])
        end
        for i = axes(data, 1)[end - 1:-1:begin]
            work[i] = fn(work[i], work[i + 1] + crnt[i])
        end
    end
    fn(work...)
end

end #module

using .Prob0082: solve_0082
export solve_0082

File: clojure/src/project_euler/sol/p0082.clj
(ns project-euler.sol.p0082
  (:require
   [clojure.string :as str]
   [project-euler.lib.util :as util]))

(defn- parse-data
  [data]
  (->> (map #(str/split % #",") data)
       (map #(map parse-long %))
       (apply map list) ; transposition
       (map #(into-array Long/TYPE %))))

(defn solve
  ([]
   (solve (util/read-data "0082_matrix.txt")))
  ([data]
   (let [matrix (parse-data data)]
     (loop [work (longs (first matrix))
            xs (next matrix)]
       (if-let [arr (longs (first xs))]
         (do (aset work 0 (+ (aget work 0) (aget arr 0)))
             (doseq [i (range 1 (alength arr))]
               (aset work i (+ (aget arr i)
                               (min (aget work (dec i)) (aget work i)))))
             (doseq [i (reverse (range (dec (alength arr))))]
               (aset work i (min (aget work i)
                                 (+ (aget arr i) (aget work (inc i))))))
             (recur work (next xs)))
         (apply min work))))))
